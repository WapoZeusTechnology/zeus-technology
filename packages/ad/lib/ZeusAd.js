import React from "react";
import PropTypes from "prop-types";
import {
  getSlotId,
  triggerPersonalizedAdPermissionsChange
} from "@zeus-technology/util";

/**
 * This component introduces a `zeus-ad` element to present an ad inventory slot.
 * @param {Object} props The properties to pass to the `zeus-ad` element.
 * @param {string} props.slotId The ID of the inventory slot.
 * @param {bool} props.debug A debug flag that prints debug messages.
 * @param {bool} props.allowPersonalization A boolean flag allowing the display of personalization in the advertisements.
 *
 * NOTE: If you use the `allowPersonalization` prop, it is best if _all_ ad slots have the same value for this. We put
 *       this prop in the `ZeusAd` component so that it can respond to change faster, but this value should not be
 *       changed more than once during any given page load, and all `ZeusAd` instances should share the same value.
 */
export class ZeusAd extends React.Component {
  constructor(props) {
    super(props);
    triggerPersonalizedAdPermissionsChange(props.allowPersonalization);

    return;
  }

  shouldComponentUpdate(nextProps) {
    this.props.debug &&
      console.log(
        `ZEUS DEBUG: ZeusAd\nReconsidering whether or not ${this.props.slotId} should be updated.`
      );

    // Handle the changing of personalization flag changes.
    if (nextProps.allowPersonalization !== this.props.allowPersonalization) {
      triggerPersonalizedAdPermissionsChange(nextProps.allowPersonalization);
    }

    return !(
      nextProps.slotId === this.props.slotId &&
      nextProps.debug === this.props.debug
    );
  }

  node() {
    return document.getElementById(getSlotId(this.props.slotId));
  }

  render() {
    const { slotId, debug, ...props } = this.props;
    const useSlotId = getSlotId(slotId);
    debug &&
      console.log(
        `ZEUS DEBUG: ZeusAd\nRe-rendering component! «${useSlotId}».`,
        JSON.parse(JSON.stringify(this.props))
      );
    return <zeus-ad class="zeus-ad" id={useSlotId} debug={debug} {...props} />;
  }
}

ZeusAd.propTypes = {
  slotId: PropTypes.string.isRequired,
  allowPersonalization: PropTypes.bool,
  debug: PropTypes.bool
};
ZeusAd.defaultProps = {
  allowPersonalization: true,
  debug: false
};
