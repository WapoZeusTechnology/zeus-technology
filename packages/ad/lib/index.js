import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";

const getSlotId = inputSlotId => `zeus_${inputSlotId.replace(/^zeus[_-]/, "")}`;

/**
 * This component introduces a `zeus-ad` element to present an ad inventory slot.
 * @param {Object} props The properties to pass to the `zeus-ad` element.
 * @param {string} props.slotId The ID of the inventory slot.
 */
export const ZeusAd = ({ slotId, ...props }) => {
  return <zeus-ad id={getSlotId(slotId)} {...props} />;
};
ZeusAd.propTypes = {
  slotId: PropTypes.string.isRequired
};

/**
 * This component introduces a `zeus-ad` element to present an ad inventory slot in a manner which responds to your Router.
 * @param {Object} props The properties to pass to the `zeus-ad` element.
 * @param {string} props.slotId The ID of the inventory slot.
 * @param {bool} props.changeOnNavigate By default, the ad slots will refresh when the route does. Pass false into this value to disable that functionality.
 */
export const ZeusAdWithRouter = withRouter(
  ({ slotId, changeOnNavigate = true, ...props }) => {
    const useSlotId = getSlotId(slotId);

    if (changeOnNavigate) {
      const element = document.getElementById(useSlotId);
      element && !!element.render && element.render();
    }

    return <ZeusAd slotId={slotId} {...props} />;
  }
);

ZeusAdWithRouter.propTypes = {
  slotId: PropTypes.string.isRequired,
  changeOnNavigate: PropTypes.bool
};
