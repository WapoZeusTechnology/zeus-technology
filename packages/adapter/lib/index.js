/**
 * Object.entries polyfill from https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries#Polyfill
 * This is because IE 11 doesn't support entries and core-js polyfill is 16kb versus this which is
 * about 30 bytes.
 */
if (!Object.entries) {
  Object.entries = function(obj) {
    var ownProps = Object.keys(obj),
      i = ownProps.length,
      resArray = new Array(i); // preallocate the Array
    while (i--) resArray[i] = [ownProps[i], obj[ownProps[i]]];

    return resArray;
  };
}

export { ZeusAdapter } from "./ZeusAdapter";
