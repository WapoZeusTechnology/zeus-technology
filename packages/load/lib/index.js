/**
 * This component performs the loading of the remote Zeus script.
 * @param {Object} props The properties passed to the component
 * @param {string} props.url The URL to load in to the script
 * @returns This function returns a `<script />` tag with the URL you provided which will work for Zeus.
 */
export const zeusLoader = ({ url }) =>
  `<script crossorigin="true" src="${url}" async="true" id="__zeus_script"></script><script type="application/javascript">document.getElementById("__zeus_script").onload=function(){window.zeus.run()}</script>`;
