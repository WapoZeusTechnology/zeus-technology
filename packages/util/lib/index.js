"use strict";

import { zeusNotice } from "./zeus-notice";
import { getSlotId } from "./getSlotId";
import {
  triggerRerender,
  triggerKeyValuePairsUpdate,
  triggerPersonalizedAdPermissionsChange,
  forceRefresh,
  forceRebuildAndRefresh
} from "./zeus-api";

export {
  zeusNotice,
  getSlotId,
  triggerRerender,
  triggerKeyValuePairsUpdate,
  triggerPersonalizedAdPermissionsChange,
  forceRefresh,
  forceRebuildAndRefresh
};
