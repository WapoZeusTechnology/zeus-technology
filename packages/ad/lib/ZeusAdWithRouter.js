import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";

import { getSlotId, zeusNotice, triggerRerender } from "@zeus-technology/util";
import {
  getNavigationFacts,
  routesAreEqual,
  routeResponderReady,
  lockZeusElement
} from "./ZeusRouteResponder";

import { ZeusAd } from "./ZeusAd";

/**
 * This component introduces a `zeus-ad` element to present an ad inventory slot in a manner which responds to your Router.
 * @param {Object} props The properties to pass to the `zeus-ad` element.
 * @param {string} props.slotId The ID of the inventory slot.
 * @param {bool} props.changeOnNavigate By default, the ad slots will refresh when the route does. Pass false into this value to disable that functionality.
 * @param {function} props.shouldChangeForRoute A function which determines whether or not we should change the ad based on the new route. This function should take two arguments: the current `location`, and the previous `location` from `react-router-dom`.
 * @param {bool} props.debug A debug flag that prints debug messages.
 */
const ZeusAdWithRouterImpl = ({
  slotId,
  location,
  shouldChangeForRoute = () => true,
  changeOnNavigate = true,
  debug = false,
  ...props
}) => {
  const navigationFacts = getNavigationFacts();
  const useSlotId = getSlotId(slotId);

  // Warn folks if there's no route responder in play.
  if (!routeResponderReady) {
    zeusNotice(
      "There is no `ZeusRouteResponder` ready. " +
        "Please add one inside your Router. See README.md for more details."
    );
  }

  // Warn the user if they gave us an invalid property here.
  if (typeof shouldChangeForRoute !== "function") {
    zeusNotice(
      "The `shouldChangeForRoute` prop passed to `ZeusAdWithRouter` is not a function!"
    );
    shouldChangeForRoute = () => true;
  }

  useEffect(() => {
    // If we navigated, then we let the caller decide whether or not we rerender.
    const shouldTriggerChange = navigationFacts.navigated
      ? shouldChangeForRoute(navigationFacts.to, navigationFacts.from) &&
        lockZeusElement(useSlotId)
      : false;
    debug &&
      console.debug(
        `ZEUS DEBUG: ZeusAdWithRouter\nConsidering a refresh of ad slot «${useSlotId}» with the following information:`,
        {
          navigationFacts,
          changeOnNavigate,
          shouldTriggerChange
        }
      );

    // Only change if we're supposed to change for this route.
    if (shouldTriggerChange && changeOnNavigate) {
      triggerRerender(useSlotId);
    }
  }, [
    slotId,
    navigationFacts,
    shouldChangeForRoute,
    debug,
    useSlotId,
    location,
    changeOnNavigate
  ]);

  return (
    <ZeusAd
      ignore={!shouldTriggerChange}
      slotId={useSlotId}
      debug={debug}
      {...props}
    />
  );
};

ZeusAdWithRouterImpl.propTypes = {
  slotId: PropTypes.string.isRequired,
  debug: PropTypes.bool,
  location: PropTypes.object.isRequired,
  changeOnNavigate: PropTypes.bool,
  shouldChangeForRoute: PropTypes.func
};

export const ZeusAdWithRouter = withRouter(ZeusAdWithRouterImpl);
