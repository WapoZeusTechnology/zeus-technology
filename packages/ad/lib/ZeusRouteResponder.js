import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import { withRouter } from "react-router-dom";
import { triggerKeyValuePairsUpdate } from "@zeus-technology/util";

export const routesAreEqual = (a, b) =>
  !!a &&
  !!b &&
  a.pathname === b.pathname &&
  a.search === b.search &&
  a.hash === b.hash &&
  a.state === b.state;

const navigationFacts = { to: null, from: null, navigated: false };
export const getNavigationFacts = () => navigationFacts;
const setNavigationFacts = ({ from, to }) => {
  // If the routes are equal, we didn't navigate!
  if (from === to || routesAreEqual(from, to)) {
    return Object.assign(navigationFacts, { navigated: false });
  }

  return Object.assign(navigationFacts, {
    navigated: true,
    from,
    to,
    timestamp: new Date().getTime()
  });
};

/**
 * This queues up something that will trigger the zeus stuff to update when KVPs change, or the route does.
 * @param {Object} props The properties to pass to the `zeus-ad` element.
 * @param {function} props.shouldChangeForRoute A function which determines whether or not we should change the ad based on the new route. This function should take two arguments: the current `location`, and the previous `location` from `react-router-dom`.
 */
export const ZeusRouteResponderImpl = ({
  location,
  keyValuePairs = {},
  shouldChangeForRoute = () => true
}) => {
  const lastLocation = useRef(location);
  const kvpRef = useRef(Object.assign({}, keyValuePairs));

  // Warn the user if they gave us an invalid property here.
  if (typeof shouldChangeForRoute !== "function") {
    zeusNotice(
      "The `shouldChangeForRoute` prop passed to `ZeusRouteResponder` is not a function!"
    );
    shouldChangeForRoute = () => true;
  }
  setNavigationFacts({ from: lastLocation.current, to: location });
  useEffect(() => {
    kvpRef.current = Object.assign({}, keyValuePairs);
    // We need to know whether or not the route changed, and if the caller thinks
    // that it is a change worthy of an update.
    const shouldTriggerChange = !routesAreEqual(lastLocation.current, location)
      ? shouldChangeForRoute(location, lastLocation.current)
      : false;
    lastLocation.current = location;

    // Only change if we're supposed to change for this route.
    if (shouldTriggerChange) {
      triggerKeyValuePairsUpdate(kvpRef.current);
    }
  }, [shouldChangeForRoute, location, keyValuePairs]);

  return null;
};

ZeusRouteResponderImpl.propTypes = {
  location: PropTypes.object.isRequired,
  keyValuePairs: PropTypes.object,
  shouldChangeForRoute: PropTypes.func
};

export const ZeusRouteResponder = withRouter(ZeusRouteResponderImpl);
