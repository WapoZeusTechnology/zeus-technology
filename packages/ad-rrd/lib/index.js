import { ZeusAd } from "@zeus-technology/ad";
import { ZeusAdWithRouter } from "./ZeusAdWithRouter";
import { ZeusRouteResponder } from "./ZeusRouteResponder";
export { ZeusAd, ZeusAdWithRouter, ZeusRouteResponder };
