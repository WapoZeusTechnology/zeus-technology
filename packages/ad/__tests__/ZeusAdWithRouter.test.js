"use strict";

import React from "react";
import { ZeusAd, ZeusAdWithRouter } from "../lib";
import renderer from "react-test-renderer";
import { HashRouter } from "react-router-dom";

describe("@zeus-platform/ad", () => {
  it("Should render and match snapshot with zeus_", () => {
    const sample = renderer.create(
      <HashRouter basename="/">
        <ZeusAd slotId="zeus_sample0" />
      </HashRouter>
    );
    expect(sample.toJSON().props.id).toEqual("zeus_sample0");
    expect(sample.toJSON()).toMatchSnapshot();
  });
});
