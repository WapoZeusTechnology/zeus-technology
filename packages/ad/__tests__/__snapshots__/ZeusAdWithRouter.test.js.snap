// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@zeus-platform/ad Should render and match snapshot with zeus_ 1`] = `
<zeus-ad
  id="zeus_sample0"
/>
`;
